File: src/main/java/AoC2023_20.java
import java.math.BigInteger;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Optional;

import com.github.pareronia.aoc.MutableInt;
import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.StringOps.StringSplit;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2023_20
        extends SolutionBase<AoC2023_20.Modules, Integer, Long> {
    
    private AoC2023_20(final boolean debug) {
        super(debug);
    }
    
    public static AoC2023_20 create() {
        return new AoC2023_20(false);
    }
    
    public static AoC2023_20 createDebug() {
        return new AoC2023_20(true);
    }
    
    @Override
    protected Modules parseInput(final List<String> inputs) {
        return Modules.fromInput(inputs);
    }
    
    @Override
    public Integer solvePart1(final Modules modules) {
        final MutableInt hi = new MutableInt(0);
        final MutableInt lo = new MutableInt(0);
        final PulseListener listener = pulse -> {
            if (pulse.isLow()) {
                lo.increment();
            } else {
                hi.increment();
            }
        };
        for (int i = 0; i < 1000; i++) {
            modules.pushButton(listener);
        }
        return hi.intValue() * lo.intValue();
    }
    
    @Override
    public Long solvePart2(final Modules modules) {
        final Map<String, List<Integer>> memo = new HashMap<>();
        final String to_rx = modules.getModuleWithOutput_rx();
        final MutableInt pushes = new MutableInt();
        final PulseListener listener = pulse -> {
            if (pulse.dest.equals(to_rx) && pulse.isHigh()) {
                memo.computeIfAbsent(pulse.src, k -> new ArrayList<>())
                    .add(pushes.intValue());
            }
        };
        for (int i = 1; i <= 10000; i++) {
            pushes.increment();
            modules.pushButton(listener);
            if (!memo.isEmpty()
                    && memo.values().stream().allMatch(v -> v.size() > 1)) {
                return memo.values().stream()
                     .map(v -> v.get(1) - v.get(0))
                     .map(BigInteger::valueOf)
                     .reduce((a, b) -> a.multiply(b).divide(a.gcd(b)))
                     .get().longValue();
            }
        }
        throw new IllegalStateException("Unsolvable");
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "32000000"),
        @Sample(method = "part1", input = TEST2, expected = "11687500"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2023_20.createDebug().run();
    }

    record Pulse(String src, String dest, Pulse.Type value) {
        public enum Type {
            HIGH, LOW;
            
            public Type flipped() {
                return this == HIGH ? LOW : HIGH;
            }
        }
        
        public static Pulse high(final String src, final String dest) {
            return new Pulse(src, dest, Type.HIGH);
        }
    
        public static Pulse low(final String src, final String dest) {
            return new Pulse(src, dest, Type.LOW);
        }
        
        public boolean isHigh() {
            return value == Type.HIGH;
        }
        
        public boolean isLow() {
            return value == Type.LOW;
        }
    }
    
    private interface PulseListener {
        void onPulse(Pulse pulse);
    }

    private static final class Module {
        private static final String BROADCASTER = "broadcaster";
        private static final String BUTTON = "button";
        private static final String SELF = "self";
        
        private final String type;
        private final List<String> outputs;
        private final Map<String, Pulse.Type> state = new HashMap<>();
        
        public Module(final String type, final List<String> outputs) {
            this.type = type;
            this.outputs = outputs;
        }

        public List<String> getOutputs() {
            return outputs;
        }

        public Map<String, Pulse.Type> getState() {
            return state;
        }
        
        private boolean isConjunction() {
            return this.type.equals("&");
        }
        
        public boolean isFlipFlop() {
            return this.type.equals("%");
        }
        
        private boolean isBroadcaster() {
            return this.type.equals(BROADCASTER);
        }
        
        public List<Pulse> process(final Pulse pulse) {
            if (this.isBroadcaster()) {
                return this.getOutputs().stream()
                    .map(o -> new Pulse(BROADCASTER, o, pulse.value))
                    .toList();
            } else if (this.isFlipFlop()) {
                if (pulse.isLow()) {
                    final Pulse.Type newState
                            = this.getState().get(Module.SELF).flipped();
                    this.getState().put(Module.SELF, newState);
                    return this.getOutputs().stream()
                        .map(o -> newState == Pulse.Type.LOW
                                    ? Pulse.low(pulse.dest, o)
                                    : Pulse.high(pulse.dest, o))
                        .toList();
                } else {
                    return List.of();
                }
            } else {
                this.getState().put(pulse.src, pulse.value());
                final boolean allHigh = this.getState().values().stream()
                                        .allMatch(v -> v == Pulse.Type.HIGH);
                return this.getOutputs().stream()
                    .map(o -> allHigh
                                ? Pulse.low(pulse.dest, o)
                                : Pulse.high(pulse.dest, o))
                    .toList();
            }
        }
    }
    
    static final class Modules {
        private final Map<String, Module> modules;

        private Modules(final Map<String, AoC2023_20.Module> modules) {
            this.modules = modules;
        }
        
        public static Modules fromInput(final List<String> inputs) {
            final Map<String, Module> modules = new HashMap<>();
            inputs.stream().forEach(line -> {
                    final StringSplit split = StringOps.splitOnce(line, " -> ");
                    final List<String> outputs = Arrays.asList(split.right().split(", "));
                    if (split.left().equals(Module.BROADCASTER)) {
                        modules.put(
                            Module.BROADCASTER,
                            new Module(Module.BROADCASTER, outputs));
                    } else {
                        final char type = split.left().charAt(0);
                        modules.put(
                            split.left().substring(1),
                            new Module(String.valueOf(type), outputs));
                    }
                });
            for (final String m : modules.keySet()) {
                final Module module = modules.get(m);
                for (final String output : module.getOutputs()) {
                    final Module m1 = modules.get(output);
                    if (m1 != null && m1.isConjunction()) {
                        m1.getState().put(m, Pulse.Type.LOW);
                    }
                }
                if (module.isFlipFlop()) {
                    module.getState().put(Module.SELF, Pulse.Type.LOW);
                }
            }
            return new Modules(modules);
        }
        
        public String getModuleWithOutput_rx() {
            return modules.entrySet().stream()
                .filter(e -> e.getValue().getOutputs().stream()
                                .anyMatch(o -> o.equals("rx")))
                .map(Entry::getKey)
                .findFirst().orElseThrow();
        }
        
        public void pushButton(final PulseListener listener) {
            final Deque<Pulse> q = new ArrayDeque<>();
            q.add(Pulse.low(Module.BUTTON, Module.BROADCASTER));
            while (!q.isEmpty()) {
                final Pulse pulse = q.pop();
                listener.onPulse(pulse);
                Optional.ofNullable(this.modules.get(pulse.dest))
                    .ifPresent(target ->
                        target.process(pulse).forEach(q::add));
            }
        }
    }

    private static final String TEST1 = """
            broadcaster -> a, b, c
            %a -> b
            %b -> c
            %c -> inv
            &inv -> a
            """;
    private static final String TEST2 = """
            broadcaster -> a
            %a -> inv, con
            &inv -> b
            %b -> con
            &con -> output
            """;
}

File: src/main/python/AoC2023_20.py
#! /usr/bin/env python3
#
# Advent of Code 2023 Day 20
#

from __future__ import annotations

import sys
from collections import defaultdict
from collections import deque
from enum import IntEnum
from functools import reduce
from math import lcm
from typing import NamedTuple
from typing import Protocol

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

TEST1 = """\
broadcaster -> a, b, c
%a -> b
%b -> c
%c -> inv
&inv -> a
"""
TEST2 = """\
broadcaster -> a
%a -> inv, con
&inv -> b
%b -> con
&con -> output
"""


class ModuleType(IntEnum):
    FLIPFLOP = 1
    CONJUNCTION = 2
    BROADCAST = 3


class PulseType(IntEnum):
    LOW = 0
    HIGH = 1

    def flipped(self) -> PulseType:
        return PulseType.LOW if self == PulseType.HIGH else PulseType.HIGH


class Pulse(NamedTuple):
    src: str
    dst: str
    value: PulseType

    @classmethod
    def low(cls, src: str, dst: str) -> Pulse:
        return Pulse(src, dst, PulseType.LOW)

    @classmethod
    def high(cls, src: str, dst: str) -> Pulse:
        return Pulse(src, dst, PulseType.HIGH)

    @property
    def is_high(self) -> bool:
        return self.value == PulseType.HIGH

    @property
    def is_low(self) -> bool:
        return self.value == PulseType.LOW


class PulseListener(Protocol):
    def on_pulse(self, pulse: Pulse) -> None:
        pass


class Module:
    BROADCASTER = "broadcaster"
    BUTTON = "button"
    SELF = "*self*"

    def __init__(self, type: str, outputs: list[str]) -> None:
        self.type = type
        self.outputs = outputs
        self.state = dict[str, PulseType]()

    @property
    def is_conjunction(self) -> bool:
        return self.type == "&"

    @property
    def is_flip_flop(self) -> bool:
        return self.type == "%"

    @property
    def is_broadcaster(self) -> bool:
        return self.type == Module.BROADCASTER

    def process(self, pulse: Pulse) -> list[Pulse]:
        if self.is_broadcaster:
            return [
                Pulse(Module.BROADCASTER, o, pulse.value) for o in self.outputs
            ]
        elif self.is_flip_flop:
            if pulse.value == PulseType.LOW:
                new_state = self.state[Module.SELF].flipped()
                self.state[Module.SELF] = new_state
                return [
                    Pulse.low(pulse.dst, o)
                    if new_state == PulseType.LOW
                    else Pulse.high(pulse.dst, o)
                    for o in self.outputs
                ]
            else:
                return []
        else:
            self.state[pulse.src] = pulse.value
            all_high = all(v == PulseType.HIGH for v in self.state.values())
            return [
                Pulse.low(pulse.dst, o)
                if all_high
                else Pulse.high(pulse.dst, o)
                for o in self.outputs
            ]


class Modules:
    def __init__(self, modules: dict[str, Module]) -> None:
        self.modules = modules

    @classmethod
    def from_input(cls, input_data: InputData) -> Modules:
        modules = dict[str, Module]()
        for line in input_data:
            a, b = line.split(" -> ")
            outputs = [_ for _ in b.split(", ")]
            if a == Module.BROADCASTER:
                modules[Module.BROADCASTER] = Module(
                    Module.BROADCASTER, outputs
                )
            else:
                modules[a[1:]] = Module(a[0], outputs)
        for m in modules:
            module = modules[m]
            for o in module.outputs:
                m1 = modules.get(o, None)
                if m1 and m1.is_conjunction:
                    m1.state[m] = PulseType.LOW
            if module.is_flip_flop:
                module.state[Module.SELF] = PulseType.LOW
        return Modules(modules)

    def get_module_with_output_rx(self) -> str:
        return next(
            k
            for k, v in self.modules.items()
            if any(o == "rx" for o in v.outputs)
        )

    def push_button(self, listener: PulseListener) -> None:
        q = deque[Pulse]()
        q.append(Pulse.low(Module.BUTTON, Module.BROADCASTER))
        while len(q) > 0:
            pulse = q.popleft()
            listener.on_pulse(pulse)
            if target := self.modules.get(pulse.dst, None):
                for p in target.process(pulse):
                    q.append(p)


Input = Modules
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return Modules.from_input(input_data)

    def part_1(self, modules: Input) -> Output1:
        class Listener(PulseListener):
            def __init__(self) -> None:
                self.lo = 0
                self.hi = 0

            def on_pulse(self, pulse: Pulse) -> None:
                if pulse.is_low:
                    self.lo += 1
                else:
                    self.hi += 1

        listener = Listener()
        for i in range(1000):
            modules.push_button(listener)
        return listener.lo * listener.hi

    def part_2(self, modules: Input) -> Output2:
        class Listener(PulseListener):
            def on_pulse(self, pulse: Pulse) -> None:
                if pulse.dst == to_rx and pulse.is_high:
                    memo[pulse.src].append(pushes)

        memo = defaultdict[str, list[int]](list)
        to_rx = modules.get_module_with_output_rx()
        pushes = 0
        listener = Listener()
        for i in range(10_000):
            pushes += 1
            modules.push_button(listener)
            if memo and all(len(v) > 1 for v in memo.values()):
                return reduce(lcm, (v[1] - v[0] for v in memo.values()))
        raise RuntimeError("unsolvable")

    @aoc_samples(
        (
            ("part_1", TEST1, 32000000),
            ("part_1", TEST2, 11687500),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2023, 20)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
